pkg/controller/watcher/watcher.go

```diff
@@ -0,0 +1,181 @@
+package watcher
+
+import (
+	"context"
+	litmuschaosv1alpha1 "github.com/litmuschaos/chaos-operator/pkg/apis/litmuschaos/v1alpha1"
+	chaosTypes "github.com/litmuschaos/chaos-operator/pkg/controller/types"
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"strings"
+)
+
+// WatchForMonitorService creates a watcher for Chaos MonitorService
+func WatchForMonitorService(client client.Client, c controller.Controller) error {
+
+	monitorServiceHandler := handlerForMonitorService(client)
+
+	err := c.Watch(&source.Kind{Type: &corev1.Service{}}, &monitorServiceHandler)
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// WatchForRunnerPod creates watcher for Chaos Runner Pod
+func WatchForRunnerPod(client client.Client, c controller.Controller) error {
+
+	runnerPodHandler := handlerForRunnerPod(client)
+
+	err := c.Watch(&source.Kind{Type: &corev1.Pod{}}, &runnerPodHandler)
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// WatchForMonitorPod creates watcher for Chaos Monitor Pod
+func WatchForMonitorPod(client client.Client, c controller.Controller) error {
+
+	monitorPodHandler := handlerForMonitorPod(client)
+
+	err := c.Watch(&source.Kind{Type: &corev1.Pod{}}, &monitorPodHandler)
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// handlerForMonitorService creates a event handler for Monitor Service
+func handlerForMonitorService(clientSet client.Client) handler.EnqueueRequestsFromMapFunc {
+	reqLogger := chaosTypes.Log.WithName("Chaos Resources Watch")
+	var monitorServiceRequest []reconcile.Request
+
+	handlerForMonitorService := handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(func(a handler.MapObject) []reconcile.Request {
+			monitorServiceCheck := strings.HasSuffix(a.Meta.GetName(), "-monitor")
+			if monitorServiceCheck {
+				svcLabels := a.Meta.GetLabels()
+				var engineUID string
+				if _, ok := svcLabels["engineUID"]; ok {
+					engineUID = svcLabels["engineUID"]
+				}
+				listOptions := []client.ListOption{
+					client.InNamespace(a.Meta.GetNamespace()),
+				}
+				var listChaosEngine litmuschaosv1alpha1.ChaosEngineList
+				err := clientSet.List(context.TODO(), &listChaosEngine, listOptions...)
+				if err != nil {
+					reqLogger.Error(err, "Unable to get the ChaosEngine Resources in namespace: %v", a.Meta.GetNamespace())
+					return nil
+				}
+
+				monitorServiceRequest = handlerRequestFromEngineList(listChaosEngine, engineUID)
```

Comment:
Line 66-81 is being repeated in below functions also.

Reply:
Yes, but the handler is different for every components, will have to keep it for now.

Reply:
We are using `handler.EnqueueRequestsFromMapFunc` for all of them. Correct me if I am wrong

Reply:
Yes, but that just stucture intialization, after that values are filled differently in each.

Reply:
Tried to refactor this again, will try to do more in upcoming PR's

Reply:
I don't see any difference in code.

Reply:
If tried to do that in single function, arguments might hinder BCH.

Reply:
You can directly pass `a` and `clientSet` to the function

Reply:
You just have to transfer lines 67-75 in a different function

Reply:
Ok, done please take a look

Reply:
Looks good